[clang] Wire -fptrauth-returns to "ptrauth-returns" fn attribute.

ahmedbougacha opened the pull request:
We already ended up with -fptrauth-returns, the feature macro, the lang opt, and the actual backend lowering.

The only part left is threading it all through PointerAuthOptions, to drive the addition of the "ptrauth-returns" attribute to generated functions.

This also adds ptrauth_key_return_address to ptrauth.h.

ahmedbougacha committed:
[clang] Wire -fptrauth-returns to "ptrauth-returns" fn attribute.

We already ended up with -fptrauth-returns, the feature macro, the lang
opt, and the actual backend lowering.

The only part left is threading it all through PointerAuthOptions, to
drive the addition of the "ptrauth-returns" attribute to generated
functions.

This also adds ptrauth_key_return_address to ptrauth.h.

kovdan01 commented:
@ahmedbougacha Non-blocking nit: you might want to add a test in clang/test/CodeGen/ptrauth-function-attributes.c (like you have for auth traps in #102417)

ahmedbougacha committed:
Restore attr codegen test.

I think it was dropped when the -fptrauth-returns flag landed in the
driver only without codegen.

ahmedbougacha committed:
Only test attrs for arm64--, only check arm64e doesn't have any by default.

Might as well while we're here.

We don't really need the triple variations anyway; one step at a time.

ahmedbougacha commented:
> @ahmedbougacha Non-blocking nit: you might want to add a test in clang/test/CodeGen/ptrauth-function-attributes.c (like you have for auth traps in https://github.com/llvm/llvm-project/pull/102417)

Oh yeah that's certainly required;  I think I dropped it accidentally when rebasing onto the driver-only part of fptrauth-returns

asl commented:
/cherry-pick 2eb6e30fe83ccce3cf01e596e73fa6385facd44b
